### PR TITLE
mining: Optionally make extra pool persistent

### PR DESCRIPTION
**Pull request by pdath**
When `rejecttokens=1`, tokens are stored in the extra pool rather than the mempool.  When Bitcoin Knots is restarted, the extra pool is lost.

The issue this creates is that compact block reconstructions for tokens become very slow after a restart because the extra pool is empty.  The causes Bitcoin Knots to have to fetch thousands of transactions from peers.  Examples are (with "debug=cmpctblock"):

`2026-06-10T06:22:16.634294Z [cmpctblock] Successfully reconstructed block 00000000000000000000aa2798940f3de06a3b6b8bf845bb872a1322a8ce6688 with 1 txn prefilled, 2181 txn from mempool (incl at least 804 from extra pool) and 2297 txn requested 2026-06-10T06:47:45.132446Z [cmpctblock] Successfully reconstructed block 000000000000000000009b2bb14d59eb073cb3aa0a1a370236b5822b99663340 with 1 txn prefilled, 2847 txn from mempool (incl at least 435 from extra pool) and 959 txn requested 2026-06-10T06:55:41.588544Z [cmpctblock] Successfully reconstructed block 00000000000000000000a51c141257c439152ffb45c6a3bded318ffaef161b78 with 1 txn prefilled, 2523 txn from mempool (incl at least 521 from extra pool) and 1800 txn requested
`

The issue does not happen when `rejecttokens=0`, because tokens then get stored in "mempool.dat", and persisted across restarts.

To resolve this issue, I propose treating the extra pool like the mempool and saving/loading the state to "extrapool.dat", controlled by a new configuration setting `persistextrapool`, that works similarly to `persistmempool`.  `persistextrapool` will default to 0 (off).

### CODE REVIEW

File: src/node/extrapool_persist.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+#include <stdexcept>
+#include <vector>
+
+using fsbridge::FopenFn;
+
+namespace node {
+
+bool ShouldPersistExtraPool(const ArgsManager& argsman)
+{
+    return argsman.GetBoolArg("-rejecttokens", false);
```
Comment (Retropex) [inline src/node/extrapool_persist.cpp]: I wonder if we should make `-persistextrapool`? Especially since your code seems to take all transactions, not just transactions affected by `-rejecttokens`.

Comment (pdath) [reply src/node/extrapool_persist.cpp]: Funny enough, I did have that at the start.
And then I reflected on the HTTP notify PR307 I made, where Luke preferred I not add an extra option, so I removed it from this PR as well.

I reflected that this option should only be used when "-rejecttokens" is used, and though it might as well trigger on that.  Miners are badly hurt if they use -rejectiontokens=1 and *don't* persist the extrapool when Bitcoin Knots is restarted.  Compact transaction reconstruction will almost certainly have to fetch large numbers of transactions from a peer without this.

I am easy on this one.

Comment (Retropex) [reply src/node/extrapool_persist.cpp]: To me it would make sense to bring it back.
Especially since `-datacarrier=0` is probably more common than `-rejecttokens=0`.

File: src/node/extrapool_persist.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        LogInfo("Failed to open extra pool file for writing: %s. Continuing anyway.\n", fs::PathToString(dump_path));
+        return false;
+    }
+
+    try {
+        // Write header: version, count, ring buffer position
+        const uint64_t version{1};
+        file << version;
+        file << count;
+        file << static_cast<uint64_t>(pool_pos);
```
Comment (kwsantiago) [inline src/node/extrapool_persist.cpp]: The dump packs out null slots (writes only non-null txns at indices 0..count-1) but still saves `pool_pos`, which indexes the original sparse array. Once eviction has left null gaps mid-buffer, the saved position no longer points at the oldest entry after reload, so ring FIFO/eviction order is wrong across restarts. Reconstruction still works (InitData scans all), but the position is misleading. Either dump the full sparse array (nulls included) or reset the saved position to `count` on load.

File: src/node/extrapool_persist.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        size_t to_load = std::min(static_cast<size_t>(count), max_count);
+        pool.resize(to_load);
+        memusage = 0;
+
+        for (size_t i = 0; i < to_load; ++i) {
+            try {
+                CTransactionRef tx;
+                file >> TX_WITH_WITNESS(tx);
+                pool[i] = std::move(tx);
+                memusage += RecursiveDynamicUsage(*pool[i]);
```
Comment (kwsantiago) [inline src/node/extrapool_persist.cpp]: Load enforces no per-tx size limit and evicts only after the whole set is deserialized. The live insert path gates on `tx_dynamic_usage < BLOCK_RECONSTRUCTION_EXTRA_TXN_PER_TXN_SIZE_LIMIT` (100000) before insert and evicts incrementally (net_processing.cpp:3059, 1813-1818). Here a corrupt/crafted `extrapool.dat` can hold oversized txns and push peak startup memory to the sum of all loaded txns before eviction trims to `max_mem_bytes`. Check per-tx `RecursiveDynamicUsage` and cumulative `memusage` inside the loop (skip/break) instead of after it.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -1699,6 +1701,24 @@ int PeerManagerImpl::GetNumberOfPeersWithValidatedDownloads() const
...
+std::pair<std::vector<CTransactionRef>, size_t> PeerManagerImpl::GetExtraPoolForDump() const
+{
+    LOCK(g_msgproc_mutex);
+    return {vExtraTxnForCompact, vExtraTxnForCompactIt};
+}
+
+void PeerManagerImpl::SetExtraPool(std::vector<CTransactionRef> pool, size_t pos, size_t memusage)
+{
+    LOCK(g_msgproc_mutex);
+    vExtraTxnForCompact.assign(m_opts.max_extra_txs, nullptr);
```
Comment (kwsantiago) [inline src/net_processing.cpp:1713]: `SetExtraPool` rebuilds capacity, position, and memusage to match the live ring-buffer invariants but has no unit coverage; the tests only exercise the free Dump/Load functions. Since the loaded pool is packed at [0,count) while `pos` comes from the old layout, the next `AddToCompactExtraTransactions` can overwrite just-loaded txns at `pos` while null slots [count,max) sit free. Add a test that calls `SetExtraPool` then `AddToCompactExtraTransactions` and checks position/memusage consistency.

File: src/node/extrapool_persist.cpp (hunk 1 above)
Comment (pdath) [reply src/node/extrapool_persist.cpp]: I have now added -persistextrapool as a config option, default off.

File: src/node/extrapool_persist.cpp (hunk 2 above)
Comment (pdath) [reply src/node/extrapool_persist.cpp]: I have completely redone this now.  pool_pos is now not saved and is calculated on load.

File: src/node/extrapool_persist.cpp (hunk 3 above)
Comment (pdath) [reply src/node/extrapool_persist.cpp]: This is now fixed.

File: src/net_processing.cpp (hunk 4 above)
Comment (pdath) [reply src/net_processing.cpp:1713]: I have now added a unit test.

**Review by Retropex — APPROVED** [review]
Concept ACK

**Comment by pdath** [comment]
I have added an extra commit to resolve a UBSAN failure in the unit test.  I have also further improved some of the error handling.  Highlights:

- Replace pre-allocation with push_back() to eliminate null entries when
  skipping oversized transactions during LoadExtraPool()
- Guard against division by zero: check max_count > 0 before modulo operation
  in pool_pos calculation
- Add test case for zero-capacity (max_count=0) handling

**Comment by rebroad** [comment]
what's the purpose of rejecttokens? why wasn't the extra pool saved alongside menpool.dat?

**Comment by pdath** [comment]
> what's the purpose of rejecttokens? why wasn't the extra pool saved alongside menpool.dat?

rejecttokens is designed to filter out tokens like BRC-20 and Runes.

You could ask "why" a lot of things are not done.  And then you can create a pull request to implement them, as I did.  :-)

**Review by Ari4ka — APPROVED** [review]

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1699,6 +1701,25 @@ int PeerManagerImpl::GetNumberOfPeersWithValidatedDownloads() const
...
+std::vector<CTransactionRef> PeerManagerImpl::GetExtraPoolForDump() const
+{
+    LOCK(g_msgproc_mutex);
+    return vExtraTxnForCompact;
+}
+
+void PeerManagerImpl::SetExtraPool(std::vector<CTransactionRef> pool, size_t memusage)
+{
+    LOCK(g_msgproc_mutex);
+    vExtraTxnForCompact.assign(m_opts.max_extra_txs, nullptr);
```
Comment (kwsantiago) [inline src/net_processing.cpp:1713]: Thanks. The functional test now covers `SetExtraPool` end-to-end (restart with 0 getblocktxn requests confirms load→reconstruction), and the load-side position derivation is well unit-tested, so the risk here is largely covered. The remaining gap is only a focused unit assertion that after `SetExtraPool` the next `AddToCompactExtraTransactions` inserts at the free slot rather than overwriting a loaded tx. Nice-to-have, not blocking.

Comment (pdath) [reply src/net_processing.cpp:1713]: I accept this case.
I tried to resolve it as part of this commit, but it got really ugly.  I then gave it further consideration and decided that, since it was an edge case and it didn't break anything, to remove the code.
